velox/vector/ComplexVector.cpp

```diff
@@ -1781,4 +1782,163 @@ void RowVector::appendNulls(vector_size_t numberOfRows) {
+  } else {
+    VELOX_USER_FAIL(
+        "Unsupported type for replacing nulls: {}",
+        TypeTraits<ValueKind>::TypeTraits::name);
+  }
+}
+
+template <TypeKind KeyKind>
+VectorPtr toRowVectorImpl(
+    const BaseVector& vector,
+    const MapVector::ToRowVectorOptions& options,
+    const SelectivityVector& rows) {
+  using KeyType = typename TypeTraits<KeyKind>::NativeType;
+  folly::F14FastMap<KeyType, size_t> keyToIndex;
+  keyToIndex.reserve(options.keysToProject->size());
+  auto keysToProjectVec = options.keysToProject->as<SimpleVector<KeyType>>();
+  for (size_t i = 0; i < keysToProjectVec->size(); ++i) {
+    VELOX_USER_CHECK(
+        !keysToProjectVec->isNullAt(i), "Keys to project cannot contain null");
+    auto key = keysToProjectVec->valueAt(i);
+    bool ok = keyToIndex.insert({key, i}).second;
+    // Key must be unique.
+    VELOX_USER_CHECK(ok, "Duplicate keys cannot be projected: {}", key);
+  }
+
+  DecodedVector decodedMap(vector, rows);
+  auto mapBase = decodedMap.base()->asUnchecked<MapVector>();
+  DecodedVector decodedKeys(*mapBase->mapKeys());
+  DecodedVector decodedValues(*mapBase->mapValues());
+  auto valueType = mapBase->mapValues()->type();
+  auto* offsets = mapBase->rawOffsets();
+  auto* sizes = mapBase->rawSizes();
+  auto outputSize = rows.end();
+
+  std::vector<VectorPtr> children;
+  children.reserve(keyToIndex.size());
+  for (size_t i = 0; i < keyToIndex.size(); ++i) {
+    children.push_back(
+        BaseVector::create(valueType, outputSize, mapBase->pool()));
+    if (!options.replaceNulls) {
+      auto rawNulls = children.back()->mutableRawNulls();
+      bits::fillBits(rawNulls, 0, outputSize, bits::kNull);
+    } else {
+      VELOX_DYNAMIC_TYPE_DISPATCH_ALL(
+          preFillDefaults, valueType->kind(), children.back());
+    }
+  }
+
+  auto outputNulls = (!options.replaceNulls && options.allowTopLevelNulls &&
+                      decodedMap.mayHaveNulls())
+      ? AlignedBuffer::allocate<bool>(
+            outputSize, mapBase->pool(), bits::kNotNull)
+      : nullptr;
+  auto* rawNulls = outputNulls ? outputNulls->asMutable<uint64_t>() : nullptr;
+  std::vector<bool> visited(keyToIndex.size());
+  std::vector<std::vector<BaseVector::CopyRange>> copyRangesFromBase(
+      keyToIndex.size());
+  rows.applyToSelected([&](vector_size_t row) INLINE_LAMBDA {
+    std::fill(visited.begin(), visited.end(), false);
+    if (!decodedMap.isNullAt(row)) {
+      auto decodedIndex = decodedMap.index(row);
+      auto offset = offsets[decodedIndex];
+      auto size = sizes[decodedIndex];
+      for (auto j = offset; j < offset + size; ++j) {
+        if (decodedKeys.isNullAt(j) || decodedValues.isNullAt(j)) {
+          continue;
+        }
+        auto key = decodedKeys.valueAt<KeyType>(j);
+        auto it = keyToIndex.find(key);
+        if (it == keyToIndex.end()) {
+          continue;
+        }
+        auto index = it->second;
+        if (visited[index]) {
+          VELOX_USER_CHECK(
```

Comment:
If this API will be used to implement a UDF, it will need to provide a no-throw behavior.

Reply:
Something like that, but there needs to be a way to distinguish rows with errors vs. null rows and allow to produce reasonable error message if errors are not suppressed.

Wondering if we would need to move away from an API on MapVector into a UDF.